Introduce a `Uri.empty` constant

diemogebhardt opened the pull request:
This PR is regarding issues https://github.com/gleam-lang/stdlib/issues/619 and https://github.com/gleam-lang/stdlib/issues/618. More specifically: https://github.com/gleam-lang/stdlib/issues/619#issuecomment-2251031871.

It introduces a `Uri.empty` constant, useful as a starting point for constructing URIs with minimal boilerplate.

## Examples

```gleam
let empty_uri = Uri.empty

// -> Uri(
//   scheme: None,
//   userinfo: None,
//   host: None,
//   port: None,
//   path: "",
//   query: None,
//   fragment: None,
// )
```

```gleam
// Create a URI by overriding only the needed fields:

let localhost_uri = Uri(..Uri.empty, scheme: Some("http"), host: Some("localhost"))

// -> Uri(
//   scheme: Some("http"),
//   userinfo: None,
//   host: Some("localhost"),
//   port: None,
//   path: "",
//   query: None,
//   fragment: None,
// )
```

diemogebhardt commented:
> Is new the right name here? What does an empty URI mean?

As s string, this constant represents `""`.  What is the proper term when picturing a `""` string, `<a href="">`, a URI/URL form field  or a browser address bar without any content?

`Uri.empty` or `Uri.blank` make sense to me as a non-native speaker. The Android API has the concept of an empty URI: https://developer.android.com/reference/android/net/Uri#EMPTY

`Uri.default`? It's not a URI containing default values.

`Uri.new`? We are not really constructing `Uri.new()` anything. But as a concept it is known and might make sense.

lpil commented:
Great, that's very helpful thank you. Let's add that 'equivalent to ""', like in the Android docs there, I think that is a lot clearer.

Could you remove one of the two examples please, it's quite verbose showing the same thing twice, and the HTML documentation and the link to the code already show the structure.

diemogebhardt commented:
Ok, done! :)

Updated one example to use the `Uri.empty`.

And changed two examples to use secure `https://example.com` instead of insecure `http://example.com`. They are just examples, yet it doesn't hurt to show best practices.

lpil commented on src/gleam/uri.gleam:
````diff
@@ -620,9 +636,9 @@ fn remove_dot_segments_loop(
 ///
 /// ```gleam
-/// let uri = Uri(Some("http"), None, Some("example.com"), ...)
+/// let uri = Uri(..Uri.empty, scheme: Some("https"), host: Some("example.com"))
````
This is a syntax error!
```suggestion
/// let uri = Uri(..empty, scheme: Some("https"), host: Some("example.com"))
```

diemogebhardt replied:
https://github.com/gleam-lang/stdlib/pull/798#discussion_r1950050996

lpil commented on src/gleam/uri.gleam:
````diff
@@ -28,6 +28,33 @@ pub type Uri {
+/// ## Examples
+///
+/// ```gleam
+/// let uri = Uri(..Uri.empty, scheme: Some("https"), host: Some("example.com"))
````
This is a syntax error!
```suggestion
/// let uri = Uri(..empty, scheme: Some("https"), host: Some("example.com"))
```

diemogebhardt replied:
These examples are being used by the LSP and on HexDocs to show the use outside of the `Uri` module, right? Isn't `Uri.empty` correct in that case then?

lpil replied:
No, that's a syntax error. Further the examples show them all locally, not from another module.

diemogebhardt replied:
Hmm – `Uri` module prefix is removed from both examples.

lpil commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@
 - Fixed a bug that would result in `list.unique` having quadratic runtime.
 - Fixed the implementation of `list.key_set` to be tail recursive.
 - The `pop` and `pop_map` functions in the `list` module have been deprecated.
+- Added `Uri.empty` constant representing an empty URI, equivalent to `""`.
```
Syntax error here! And this isn't the convention used in this repo.
```suggestion
- The `uri` module gains the `empty` value, representing an empty URI which 
  equivalent to `""`.
```

lpil reviewed (approved):
Thank you!
